File: lda/ruby/lda.rb
def lda(data, n, d)
    # Initialize the coef and means arrays.
    coef = Array.new(d, 0)
    means = Array.new(d, 0)
  
    # Calculate the means of the two classes.
    for i in 0...n
      if data[i * d] == 0
        means[0] += data[i * d]
      else
        means[1] += data[i * d]
      end
    end
    means[0] /= n / 2.0
    means[1] /= n / 2.0
  
    # Calculate the coef array.
    for i in 0...d
      for j in 0...n
        if data[j * d] == 0
          coef[i] += data[j * d] - means[0]
        else
          coef[i] += data[j * d] - means[1]
        end
      end
      coef[i] /= n / 2.0
    end
  
    coef
  end
  
  # Initialize the data array.
  data = [1.0, 2.0, 0.0, 3.0, 4.0, 5.0, 1.0, 2.0, 1.0]
  n = 3
  d = 3
  
  # Calculate the coef array.
  coef = lda(data, n, d)
  
  # Print the coef array.
  p coef
  

File: lda/kotlin/lda.kt
fun LDA(data: FloatArray, n: Int, d: Int, coef: FloatArray, means: FloatArray) {
    // Initialize the coef and means arrays.
    coef.fill(0.0f)
    means.fill(0.0f)

    // Calculate the means of the two classes.
    for (i in 0 until n) {
        if (data[i * d] == 0f) {
            means[0] += data[i * d]
        } else {
            means[1] += data[i * d]
        }
    }
    means[0] /= n / 2.0f
    means[1] /= n / 2.0f

    // Calculate the coef array.
    for (i in 0 until d) {
        for (j in 0 until n) {
            if (data[j * d] == 0f) {
                coef[i] += data[j * d] - means[0]
            } else {
                coef[i] += data[j * d] - means[1]
            }
        }
        coef[i] /= n / 2.0f
    }
}

fun main(args: Array<String>) {
    // Initialize the data array.
    val data = floatArrayOf(1.0f, 2.0f, 0.0f, 3.0f, 4.0f, 5.0f, 1.0f, 2.0f, 1.0f)
    val n = 3
    val d = 3

    // Initialize the coef and means arrays.
    val coef = FloatArray(d)
    val means = FloatArray(d)

    // Calculate the coef and means arrays.
    LDA(data, n, d, coef, means)

    // Print the coef array.
    for (i in 0 until d) {
        println(coef[i])
    }
}
